File: backend/internal/domain/review.go
package domain

import "time"

// Review is a guest's rating + comment on a listing. It maps onto the legacy
// `review` table, whose `owner_id` column holds the review's author (the user
// who wrote it) and whose `house_id` points at the reviewed listing. The
// `created_at` column is a mobile addition (see migration 000003).
type Review struct {
	ID              int32
	HouseID         int32
	AuthorID        int32
	AuthorName      string
	AuthorAvatarURL string
	Rating          int32
	Body            string
	Status          string
	RejectionReason string
	RequestID       *int32
	CreatedAt       time.Time
	Reply           *ReviewReply

	// Optional house metadata, populated for my reviews pages
	HouseStreet    string
	HouseNumber    string
	HouseCity      string
	HouseCoverPath string
}

type ReviewReply struct {
	ID              int64
	ReviewID        int32
	OwnerID         int32
	Body            string
	Status          string
	RejectionReason string
	CreatedAt       time.Time
}

type ReviewEligibility struct {
	RequestID       int32     `json:"request_id"`
	HouseID         int32     `json:"house_id"`
	CanReview       bool      `json:"can_review"`
	ReviewDeadline  time.Time `json:"review_deadline"`
	ReviewID        *int32    `json:"review_id,omitempty"`
	ReviewStatus    string    `json:"review_status,omitempty"`
	ReviewRating    *int32    `json:"review_rating,omitempty"`
	ReviewBody      string    `json:"review_body,omitempty"`
	RejectionReason string    `json:"rejection_reason,omitempty"`
	EditAttempts    int32     `json:"edit_attempts,omitempty"`
	MaxAttempts     int32     `json:"max_attempts,omitempty"`
}

// RatingSummary aggregates a listing's published reviews. Distribution is
// indexed by stars-1, so Distribution[0] is the number of 1-star reviews and
// Distribution[4] the number of 5-star reviews.
type RatingSummary struct {
	Average      float64
	Total        int32
	Distribution [5]int32
}

// NewReview carries the validated fields needed to create a review.
type NewReview struct {
	RequestID int32
	HouseID   int32
	AuthorID  int32
	Rating    int32
	Body      string
}

type ReviewModerationJob struct {
	ID         int64
	TargetType string
	TargetID   int64
	Attempts   int32
}

type ReviewModerationTarget struct {
	TargetType     string
	TargetID       int64
	ReviewID       int32
	HouseID        int32
	AuthorID       int32
	ReviewAuthorID int32
	Rating         int32
	Body           string
	MaskedBody     string
	Categories     []string
}

File: mobile/src/types/review.ts
/**
 * Review types mirror the backend DTOs (snake_case kept 1:1 with the wire
 * format). See backend/internal/delivery/http/review_handler.go.
 */

export interface Review {
  id: number;
  /** 1..5 stars. */
  rating: number;
  body: string;
  author_name: string;
  /** Absolute avatar URL, or '' when the author has none. */
  author_avatar_url: string;
  /** RFC3339 timestamp. */
  created_at: string;
  status?: 'pending_moderation' | 'active' | 'rejected' | 'moderation_review';
  rejection_reason?: string;
  reply?: ReviewReply;
}

export interface ReviewReply { id:number; body:string; status?:string; rejection_reason?:string; created_at:string }
export interface ReviewEligibility { request_id:number; house_id:number; can_review:boolean; review_deadline:string; review_id?:number; review_status?:string; review_rating?:number; review_body?:string; rejection_reason?:string; edit_attempts?:number; max_attempts?:number }

export interface ReviewSummary {
  /** Average score rounded to one decimal (0 when there are no reviews). */
  average: number;
  total: number;
  /** Count of reviews per star value, keyed by "1".."5". */
  distribution: Record<string, number>;
}

export interface ReviewsPage {
  summary: ReviewSummary;
  items: Review[];
  total: number;
  limit: number;
  offset: number;
}

/** Body of POST /api/v1/requests/{id}/review. */
export interface CreateReviewBody {
  rating: number;
  body: string;
}

export interface UserReview {
  id: number;
  rating: number;
  body: string;
  author_name?: string;
  author_avatar_url?: string;
  created_at: string;
  house_id: number;
  house_street: string;
  house_number: string;
  house_city: string;
  house_cover_url: string;
  status?: string;
  rejection_reason?: string;
  reply?: ReviewReply;
  request_id?: number;
}

export interface UserReviewsPage {
  items: UserReview[];
  total: number;
  limit: number;
  offset: number;
}

